ARCHITECTURE.md

```diff
@@ -0,0 +1,209 @@
+  * A good general principle for code.org: given our problem space, we should usually be making pretty boring technology choices.
+
+</details>
+
+<details>
+  <summary>Exceptions</summary>
+
+  * When it is fundamentally impossible to implement a specified feature without using another language, e.g. Javabuilder uses Java because our curriculum goal required running Java code, which inherently requires Java.
+  * If it is not feasible to implement a feature in Ruby due to lack of libraries, Python may be invoked from ActiveJob ([howto](https://github.com/code-dot-org/code-dot-org/tree/staging/python)).
+  * May use *short* snippets of inline JS/JQuery *without* React on the client side, where pragmatic.
+  * May use bash or nodejs for general tooling when pragmatic (e.g. better nodejs apis than ruby)
+  * Consider Redis instead of MySQL when performance is important and ephemerality is acceptable (e.g. cache values)
+  * This tenet will probably have more exceptions and edge cases than most. Its not an exhaustive list, its a "highly preferred list".
+    * Be pragmatic, do a good faith attempt to use preferred languages and frameworks, but solve your problems.
+    * E.g. we also use cloudformation (etc etc) which could be argued to be kind of like a programming language, but is **totally** not a problem.
+      * OTOH, you don't get to "use Haskell because everything else should be considered lame and fundamentally broken" 😉.
+
+</details>
+
+### Tenet 5: “The Rails Way” (vs. overfitting to today’s needs)
+
+Prefer the conventional Rails way to accomplish a task: it might be better to be 100% conventional than get a 10% gain by deviating from convention. This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.
+
+General thoughts on the Rails approach can be found at https://rubyonrails.org/doctrine with more specific advice for "how do to X the rails way" at https://guides.rubyonrails.org/. If you're not familiar with Rails, a course or a book might be helpful.
+
+<details>
+  <summary>Why?</summary>
+
+  * Observing the codebase over a decade of dev: sometimes you can get a gain this year by customizing Rails to work a different way that’s a little better for us, but over time non-conventional code tends to become a maintenance burden.  
+  * Often the “10% better” non-conventional code is exactly what prevents us from upgrading to the “50% better” official Rails solution to the problem that comes out 2 years later.  
+  * By and large we have pretty conventionally Rails-shaped problems, which means Rails tends to solve our problems better than we would if we wait a little longer.  
+  * Rails has been a great long-term fit to our problem space and we benefit from continuing to get upgrades as Rails has improved. Doing it “their way” makes upgrades happen semi-automatically.
```

Comment:
there are some ways we may want to deviate from the rails due to the size of our application:
* we may want to modularize our rails monolith into separate groups of business logic which own their own data, using rails engines (see [Domain Boundaries - A Starting Point](https://docs.google.com/document/d/1WffY2sr3f-dJQwqj98fhCQhqXhf3bmiaehKlpw21BBc/edit) (2021)). this is not part of the rails way, but may be advantageous to us so that we can modularize our monolith into separate units of business logic which each own their own data.
* rails steers us toward routes which are tied to models, however we should consider that routes which represent higher-level abstractions (e.g. teacher dashboard) should not be associated with individual models. 

would an exception like this make sense to add here:
* there may be some ways where deviating from the rails way makes sense due to the large size of our application, such as by modularizing our rails app via rails engines or by decoupling rails routes from being associated with individual models.

my thinking is that the emphasis on our application size gives a more general sense of when we might consider deviating, rather than listing the more specific bullets as the exceptions themselves.

Reply:
I like the observation that many deviations from default rails will be driven by the relative size of our app. That said, we're linking to https://guides.rubyonrails.org/ as a first blush "rails way" and we have: https://guides.rubyonrails.org/engines.html, so I'm not sure I'd think of engines as deviating from the rails way?

Reply:
I've added this exception:
```
<details>
  <summary>Exceptions</summary>

    * Sometimes deviating from stock Rails makes sense due to the large size of our application:
      * e.g. modularizing our rails app via [Rails engines](https://guides.rubyonrails.org/engines.html).
      * e.g. decoupling rails routes from being associated with individual models.

</details>
```

Reply:
oops, nice catch! I think it's still helpful to include, so thanks for adding.